### PR TITLE
HPCC-21267 Possible issue in Dali regarding timestamp comparison

### PR DESCRIPTION
**Pull request by RussWhitehead**
When Dali checks the request timestamp to see if it is expired or from the
future, it should ignore nanoseconds in order to eliminate occasional
unexpected failures. Also, in the case of a failure, add logging of the given
request time stamp and the actual Dali time stamp, both in UTC, which should
help debug failures.

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@rpastrana Please review

File: dali/server/daldap.cpp [hunk 1]
```diff
@@ -164,19 +164,23 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
 
                 CDateTime now;
                 now.setNow();
-                if (now.compare(reqUTCTimestamp) < 0)//timestamp from the future?
+                if (now.compare(reqUTCTimestamp, false) < 0)//timestamp from the future?
```
Comment (rpastrana) [inline dali/server/daldap.cpp]: How do we ensure the req timestamp is  in UTC?

File: dali/server/daldap.cpp [hunk 2]
```diff
@@ -164,19 +164,23 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
...
                 {
-                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Request digital signature timestamp %s from the future",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str());
+                    StringBuffer localDaliTimeUTC;
+                    now.getString(localDaliTimeUTC, false);//get UTC timestamp
+                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Request digital signature UTC timestamp %s from the future (Dali UTC time %s)",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str(), localDaliTimeUTC.str());
```
Comment (rpastrana) [inline dali/server/daldap.cpp]: Is there any worthwhile guidance we can provide to the admin reading the log regarding these types of errors?
-- confirm clock settings on XYZ machine, etc?

**Review by rpastrana — APPROVED** [review]
@RussWhitehead looks reasonable, just a couple of questions.

File: dali/server/daldap.cpp (hunk 1 above)
Comment (RussWhitehead) [reply dali/server/daldap.cpp]: it is set by the caller (runtime engine), in dasess.cpp method createDaliSignature, which does this
        StringBuffer timeStr;
        now.setNow();
        now.getString(timeStr, false);//get UTC timestamp

File: dali/server/daldap.cpp (hunk 2 above)
Comment (RussWhitehead) [reply dali/server/daldap.cpp]: Perhaps, but it should be obvious !

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
@rpastrana Is this approved now?

**Review by rpastrana — APPROVED** [review]
@RussWhitehead approved

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by RussWhitehead** [comment]
@richardkchapman The Vault team is asking for this.  If there is another 7.0.6 release coming out let me know and I can rebase

**Pull request merged**